src/registry.rs

```diff
@@ -0,0 +1,164 @@
+    next: Atomic<Entry>,
+}
+
+/// Registers a thread by inserting a new entry into the list of participating threads.
+///
+/// Returns a pointer to the newly allocated and inserted entry.
+fn register() -> *const Entry {
+    let list = participants();
+
+    let mut new = Owned::new(Entry {
+        participant: Participant::new(),
+        next: Atomic::null(),
+    });
+
+    unsafe {
+        // Since we don't dereference any pointers in this block, it's okay to use `unprotected`.
+        unprotected(|scope| {
+            let mut head = list.load(Acquire, scope);
+
+            loop {
+                new.next.store(head, Relaxed);
+
+                // Try installing the new entry as the new head.
+                match list.compare_and_set_weak_owned(head, new, AcqRel, scope) {
+                    Ok(n) => return n.as_raw(),
+                    Err((h, n)) => {
+                        head = h;
+                        new = n;
+                    }
+                }
+            }
+        })
+    }
+}
+
+thread_local! {
+    /// The thread registration harness.
+    ///
+    /// The harness is lazily initialized on its first use, thus registrating the current thread.
+    /// If initialized, the harness will get destructed on thread exit, which in turn unregisters
+    /// the thread.
+    static HARNESS: Harness = Harness {
+        entry: register(),
+    };
+}
+
+/// Holds a registered entry and unregisters it when dropped.
+struct Harness {
+    entry: *const Entry,
+}
+
+impl Drop for Harness {
+    fn drop(&mut self) {
+        unsafe {
+            let entry = &*self.entry;
+
+            // Unregister the thread by marking this entry as deleted.
+            unprotected(|scope| { entry.next.fetch_or(1, Release, scope); });
+        }
+    }
+}
+
+/// Returns a reference to the head pointer of the list of participating threads.
+fn participants() -> &'static Atomic<Entry> {
+    static PARTICIPANTS: AtomicUsize = ATOMIC_USIZE_INIT;
+    unsafe { &*(&PARTICIPANTS as *const AtomicUsize as *const Atomic<Entry>) }
+}
+
+/// Acquires a reference to the current participant.
+///
+/// Participants are lazily initialized on the first use.
+///
+/// # Panics
+///
+/// If this function is called while the thread is exiting, it might panic because it accesses
+/// thread-local data.
+pub fn with_current<F, R>(f: F) -> R
+where
+    F: FnOnce(&Participant) -> R,
+{
+    HARNESS.with(|harness| {
+        let entry = unsafe { &*harness.entry };
+        f(&entry.participant)
+    })
+}
+
+/// Returns an iterator over all participating threads.
+///
+/// Note that the iterator might return the same participant multiple times.
+pub fn iter(scope: &Scope) -> Iter {
+    let pred = participants();
+    let curr = pred.load(Acquire, scope);
+    Iter { scope, pred, curr }
+}
+
+/// An iterator over all participating threads.
+///
+/// Note that the iterator might return the same participant multiple times.
```

Comment:
As far as I understood, this iterator behaves like RCU linked lists, in the sense that insertion, removal, and traversal are not linearizable.  For example, `Harness:drop()` is performing `fetch_or` on `Entry::next` with `Release` ordering, but a concurrent traversal may ignore it.

I think it would be nice if we could comment more on this data structure.  What is the specification?  and Why it works?